### PR TITLE
UNITY: Support building and testing WEB in CMS CI

### PR DESCRIPTION
**Pull request by jonpugh**
## Original Issue

https://va-gov.atlassian.net/browse/VAGOV-5481  

Hit up @ElijahLynn for access.

## Stories

This is all about "CMS breaks web builds". We discovered it's mostly because WEB has fallbacks and failsafes so that it will build without the CMS at all.

As a CMS developer,
I need the WEB project to build automatically and fail transparently when the CMS breaks something
So that I don't unintentionally break the WEB build.

As a CMS developer,
I need the WEB project to build automatically from the CMS instance I am developing
So that I can be sure the WEB project can consume the content without breaking.

We need to disable all failsafes and all backup content so that developers AND content editors can be sure their content is making it to the front-end build.

## Description

This PR contains the changes needed in WEB to allow us to test the CMS breaking WEB builds.

## Changes Needed

- [x] Extra repos needed: https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started  (@jonpugh could not get it building at all without the `vagov-content` repo present. I have no preference how this happens, we just want `yarn install` to pass correctly and exit with a proper exit code.)
- [x] Add `--ignore-failsafe` (or similar) to allow code to exit at different places like:
  - [x] Drupal site unreachable, with or without proxy.
  - [x] Drupal site gives error.
  - [x] Graph QL gives error.
  - [x] GraphQL is incompatible.
  - [x] "Failed to pipe Drupal content into Metalsmith" ... Do all builds that can't get Drupal content say this? MIght be possible to only check for failsafes here.
  - [x] Exit immediately (with non-zero code) when these errors occur.
- [x] Add `--pull-drupal` to our `composer.json` so that web build always pulls from drupal cache.
- [x] Output Drupal URL at the bottom, next to "Successfully piped Drupal content into Metalsmith!"
- [ ] Discover what else is needed.

## CMS-Native Preview

Now that we are including WEB source code, we can do builds and previews directly in CMS for editors.

### "Single Page Preview" vs "Fast Site Rebuild"

"Single Page Preview" is very problematic, as each "page" is actually composed of multiple pieces of content. 

I recommend we deprecate "Single Page Preview" in favor of "Fast Site Rebuilds". If static site rebuilds take less than 1 minute, and are triggered the moment an Editor hits "Save", the cognitive delay will be minimal, and the payoff is a much more accurate, full site preview. 

This would allow editors to see their content as it would appear on *other* pages, for example, when a News article is listed on the homepage, an editor should be able to preview their changes on the main article page, *and* the homepage. Full Site Preview would allow this.

## Testing done

See https://github.com/department-of-veterans-affairs/va.gov-cms/pull/391

## Screenshots
@TODO

## Acceptance criteria
- [ ] Entire WEB app can be installed in one command, including all repo dependencies.

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Disable proxy by default, ensuring proxy is still enabled wherever VSP team expects it to be. (Still needed: See https://github.com/department-of-veterans-affairs/va.gov-cms/pull/595#commitcomment-35199205)

### CODE REVIEW

File: script/install-repos.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
+git clone git@github.com:department-of-veterans-affairs/vets-website.git
```
Comment (jbalboni) [inline script/install-repos.sh]: Wouldn't this one already be cloned if you were running this script?

File: script/install-repos.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+git clone git@github.com:department-of-veterans-affairs/vagov-content.git
```
Comment (jbalboni) [inline script/install-repos.sh]: These are being cloned to a location inside vets-website, but they should be cloned as siblings to vets-website. 

vagov-content is the only one that is required for the build process, it's possible we're trying to do too much by installing all of them.

File: package.json [hunk 3]
```diff
@@ -23,6 +23,7 @@
     "nightwatch": "nightwatch -c config/nightwatch.js",
     "nightwatch-sauce": "nightwatch -c config/nightwatch-sauce.js",
     "nightwatch-visual": "node src/platform/testing/visual-regression/index.js",
+    "install": "./script/install-repos.sh",
```
Comment (jbalboni) [inline package.json]: One thing to note: developers run `yarn install` regularly, because it brings your dependencies up to date if there have been changes. Looks like this script fails if the repos are already cloned.

File: script/install-repos.sh (hunk 1 above)
Comment (jonpugh) [reply script/install-repos.sh]: Yes, that was a late night copy-paste error :)

File: package.json (hunk 3 above)
Comment (jonpugh) [reply package.json]: Ok, no problem... I'll change the script to not fail if dirs are already present

Comment (rianfowler) [reply package.json]: I think we should rename this and document this behavior in the readme as a separate command. Cloning repos outside of package directory is not generally expected behavior for an install script. I think the install script should _only_ install  _this_ package's dependencies into `node_modules`.

We should also make sure that any changes to the install include changes to the documentation: https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/

File: script/install-repos.sh (hunk 2 above)
Comment (jonpugh) [reply script/install-repos.sh]: Ok. I've changed the script to clone to parent dir, and commented out the other repos until we confirm they are not needed.

**Comment by jonpugh** [comment]
@jbalboni Is the code that clones the repos contained in the Jenkins config somewhere?

Can you point it out to me? 

Thanks

**Comment by ncksllvn** [comment]
@jonpugh Can you expand on what it means to unify the build process for CMS & WEB?

**Comment by jonpugh** [comment]
A technical breakdown is in the description of this PR: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/391

Basically, PHP's `composer` project, (analogous to `npm` in the node world) can now do a number of cool things that, when used together, allow us to build and launch both CMS and WEB from a single repo:

- The `mouf/nodejs-installer` package downloads and installs `node` and `npm` as an executable in the CMS `/bin` dir, right next to other tools.
- This package lets us pin the version of node and NPM, making it easy to keep the build stable. 
- Composer's "repositories" feature allows us to install the entire `vets-website` codebase into our vendor directory.
- Composer's `scripts`, (exactly like packages.json scripts) allow us to kick off the commands needed to install `vets-website`... See the `va:web:install` command in our composer.json: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/391/files#diff-b5d0ee8c97c7abd7e3fa29b9a27d1780R283
  This means the same command we must use to build the CMS (`composer install`) now also builds the WEB app in the same place.
- PHP has a built in web server just like node, so we are leveraging that to create a simplified way to stand up a development environment, and we created a command to run the `npm watch` command that is used by the WEB project:
  ```
  $ composer va:start   # Installs CMS using SQLite, then launches a simple web server
  $ composer va:web:start   # Runs `npm watch`
  ```
- I am creating a VA CMS Install Profile so we can use that one command to create a fully function CMS instance without a database copy.
- By making WEB a dependency of CMS using composer, we can pin releases of CMS to a version of WEB. This means we could have a single release process, but still maintain different versions of WEB and CMS. The `composer install` process could generate release artifacts for both CMS and WEB at once.
- By making the PHP server work simply, we can give WEB developers an easy way to launch an instance of CMS for testing against.
- We can unify CI, and just have one pipeline that installs both projects and runs both projects tests.

File: package.json (hunk 3 above)
Comment (cvalarida) [reply package.json]: I'm with @rianfowler. If I run `yarn install`, this not only updates the dependencies, but runs this scrip to install the repos. Cloning repos outside of the working directory as a side-effect of updating my dependencies isn't great.

What do you think about changing this to `install-repos`? Can you update your `composer` build script to run both `install` and `install-repos`? (I remember basically nothing about `composer` from my PHP days. :sweat_smile:)

**Comment by cvalarida** [comment]
Cool idea!

I'm not keen on opening up an avenue to accidentally clone repositories outside of the working directory when trying to update my dependencies, but I'm hoping that can be fixed pretty easily. :crossed_fingers:

File: composer.lock [hunk 4]
```diff
@@ -0,0 +1,140 @@
+{
```
Comment (cvalarida) [inline composer.lock]: Naive question: Does this _have_ to be in here? I think I understand what you're trying to do with it, but having a `composer.lock` file in a front end, JavaScript-centric repository makes me squirm a bit.

File: package.json (hunk 3 above)
Comment (jonpugh) [reply package.json]: > Can you update your composer build script to run both install and install-repos?

Absolutely, composer scripts are arrays, I can add extra commands.

I just thought that, since WEB does depend on this repo to work, it would be better to include it in the WEB repo. But if you think differently, I can work around it, no problem.

File: composer.lock (hunk 4 above)
Comment (jonpugh) [reply composer.lock]: Now you know how we feel... ;)

https://github.com/department-of-veterans-affairs/va.gov-cms/blob/develop/package.json

Not sure yet. The .lock file could probably be ignored.

I'll keep trying to set up a pseudo-repo on our end so we don't need a composer.json on your end either.

**Comment by jonpugh** [comment]
# An Update!


1. I removed the `composer.json` files for now. We don't need them anymore. However, this means we have to manually update the SHA of `vets-website` in our `composer.json` right now. Solutions:
  1. Add back `composer.json` and submit to packagist. Then, packagist.org would respond to new releases, and we would only have to run `composer update va-gov/web` to get our codebase to have the latest release of `vets-website`.
  2. Host a private packagist? People have done it.
2. We have MERGED the PR in CMS that loads `vets-website` at the commit 4c8a42a43aad335adba2a7e7fc865804cf516d2a. Our CMS CI now provides `vets-website` built from the CMS PR instance (almost...)
3. We can't quite get `vets-website` to properly fail if it cannot get the Drupal graphql or reach the site. The build passes, so we don't know there are errors. 

The changes that are needed above are required for us to implement full CI on WEB+CMS with end-to-end testing, and to stabilize the content editing and previewing process.

We're looking forward to working together to get this done!

Thanks,
Jon

**Comment by jonpugh** [comment]
The merged PR is here: https://github.com/department-of-veterans-affairs/va.gov-cms/pull/493

@ElijahLynn will be submitting another PR to CMS with the latest SHA of this  project so we can get started on failing the build!

**Comment by cvalarida** [comment]
I _think_ I understand the purpose behind this PR, but I'm not seeing an originating ticket anywhere in the conversation. Can you provide that link, @jonbot?

File: package.json (hunk 3 above)
Comment (cvalarida) [reply package.json]: Is this still needed?

File: src/site/stages/build/drupal/api.js [hunk 5]
```diff
@@ -37,10 +38,22 @@ function getDrupalClient(buildOptions) {
...
+    // CMS using the AWS ELB URL.
+    // e.g. internal-dsva-vagov-prod-cms-2000800896.us-gov-west-1.elb.amazonaws.com
+
+    // When running WEB build process, the site at DRUPAL_ADDRESS must be reachable.
+    // If building and sourcing content from a site at *.cms.va.gov, from outside of VAEC,
+    // the socks proxy agent must be used because those sites are in VAEC.
+
+    // If a CMS is available at DRUPAL_ADDRESS, reachable by the WEB build process, the proxy is not needed.
+    // This includes in local development environment where user has a local CMS instance
+    // and when WEB is built in CMS-CI where every WEB has a CMS instance.
```
Comment (cvalarida) [inline src/site/stages/build/drupal/api.js]: This is a lot to digest. While I appreciate the level of detail, I'm not sure I understand it all.

My main concern is usability; if I tell the build script to `--use-proxy` when I shouldn't or don't `--use-proxy` when I should, is it going to error out? If so, does that error make it clear what I should do to fix it?

If it's user-friendly, then having this level of detail in here makes enough sense. If it isn't, I'll have to hunt through the code to figure out what the story is, and when I run into this, it may not be clear to me:
1) That this comment block is related to my problem
2) What I should do to fix it

**Review by cvalarida — COMMENTED** [review]
I'm still wary of the unexpected side-effect of doing a `yarn install`.

**Comment by cvalarida** [comment]
:man_facepalming: Definitely did. Sorry about that!

File: src/site/stages/build/drupal/metalsmith-drupal.js [hunk 6]
```diff
@@ -102,12 +102,14 @@ async function loadDrupal(buildOptions) {
...
-    log('Drupal content unavailable in cache');
+    log('Drupal content unavailable in local cache: ' + drupalCache);
     shouldPullDrupal = true;
+  } else {
+    log('Drupal content loaded from local cache: ' + drupalCache);
   }
 
   if (shouldPullDrupal) {
-    log('Attempting to load Drupal content from API...');
+    log(`Attempting to load Drupal content from API at: ${contentApi.getSiteUri()}`);
```
Comment (ElijahLynn) [inline src/site/stages/build/drupal/metalsmith-drupal.js]: @cvalarida This would be good to pick into a separate PR.

**Comment by ElijahLynn** [comment]
Crosslinking related ticket.  https://github.com/department-of-veterans-affairs/va.gov-team/issues/951

File: package.json (hunk 3 above)
Comment (jonpugh) [reply package.json]: Only one way to find out :)  ...

Does `yarn install` and `npm build` work if the content repo is not there?

File: src/site/stages/build/drupal/api.js (hunk 5 above)
Comment (jonpugh) [reply src/site/stages/build/drupal/api.js]: I think the confusion for us comes from the fact that devops/infrastructure/application boundaries are vague here.

This proxy is setup using the application code, and the URLs are hard coded for the environment.

In our world, we design the app to assume the connection is available, and setup the infrastructure (or local development tools) to setup the proxy if needed.

**Review by jonpugh — COMMENTED** [review]
I opened this as a review because it's a big discussion that is about high level architecture: How much infrastructure/environment specific information is coded into the application?

From my perspective as a CI manager, I would prefer the application code didn't worry about proxies at all to reduce the number of variables at play when trying to get things to work.

I would prefer to remove all of the proxy connection things, from the CMS-CI perspective. 

We run a  CMS and a WEB instance right next to each other in CMS-CI. The logic checking for proxy actually broke our setup when we first got it going because it assumed it was needed and we couldn't figure out a way to turn it off.

Let's prioritize a discussion on this. We need to resolve it in the next 10 days.

File: package.json (hunk 3 above)
Comment (jonpugh) [reply package.json]: Ok, after chatting with @cvalarida the content repo is still required, but they still don't want to use the `yarn install` command to do it.

So, we will: 

- [ ] Create a new `yarn install-content` command.
- [ ] Remove the Shell script.
- [ ] Add a call to `yarn install-content` before `yarn install` in our build process in `composer.json`

File: src/site/stages/build/drupal/api.js (hunk 5 above)
Comment (jonpugh) [reply src/site/stages/build/drupal/api.js]: @cvalarida Big picture aside, do you like the useProxy option? We can set whatever default you want it to have. We need it for CI so having a CLI option is fine for us.

If I update the comment here to be more descriptive, would the useProxy parts be mergable?

File: package.json (hunk 3 above)
Comment (cvalarida) [reply package.json]: @jonpugh Sounds about right! Thanks for this! :raised_hands:

Comment (jonpugh) [reply package.json]: Actually, @cvalarida would you be able to handle that todo list for us?

I'm not personally versed in Yarn yet, I feel like this would take you about 30 seconds...  :)

LMK!

File: src/site/stages/build/drupal/api.js (hunk 5 above)
Comment (jonpugh) [reply src/site/stages/build/drupal/api.js]: Ok, @cvalarida I tested out WEB master branch and we do need to figure out the Proxy defaults.

In our CI system, it defaults to enabled. See the web build failure here:
> FetchError: request to http://pr595.ci.cms.va.gov/flags_list failed, reason: connect ECONNREFUSED 127.0.0.1:2001


https://github.com/department-of-veterans-affairs/va.gov-cms/pull/595#commitcomment-35199205

Comment (cvalarida) [reply src/site/stages/build/drupal/api.js]: I added a `--no-drupal-proxy` flag a couple days ago. Does that solve this problem?

**Comment by ElijahLynn** [comment]
I just reset this branch back a few commits to 6ffa26b1089284e814662ca9b76ab5bdbf2bbd38 to get our current CMS master passing. 

image

... And I am closing this branch as all the functionality we needed has been merged in these PRs in smaller, refactored chunks:

1. https://github.com/department-of-veterans-affairs/vets-website/pull/10832 (install-repos script)
2. https://github.com/department-of-veterans-affairs/vets-website/pull/10803 (composer.json/packagist)
3. https://github.com/department-of-veterans-affairs/vets-website/pull/10763 (--drupal-fail-fast)
4. https://github.com/department-of-veterans-affairs/vets-website/pull/10799 (--no-drupal-proxy)
